### DISCUSSION TITLE
Real-time collaborative editing with editor.js

### DISCUSSION THREAD

**Discussion by teilzeitgeist** [codex-team/editor.js#1874] [Ideas]
Hey!

Did someone test to get editor.js work with SwellRT or other solutions to make the editing collaborative?

**Comment by oliver-dvorski** [comment]
You'd still need to write quite a bit of logic to make these two communicate, no?

**Comment by milu2003** [comment]
i need it too!

**Comment by marcoancona** [comment]
I tried to get real-time collaboration with Firebase Real-time Database but editor.js lacks fundamental APIs to get this to work nicely.

In particular, without a mechanism to render only some blocks rather than all, a mechanism to pass metadata to blocks or to make individual blocks read-only, it is very hard.

**Comment by milu2003** [comment]
> I tried to get real-time collaboration with Firebase Real-time Database but editor.js lacks fundamental APIs to get this to work nicely.
> 
> In particular, without a mechanism to render only some blocks rather than all, a mechanism to pass metadata to blocks or to make individual blocks read-only, it is very hard.

editorJS didn't think of this from the beginning of design. Multi person real-time collaboration is not only a problem of local rendering, such as how to undo multi person real-time collaboration?

How to compare the differences between the two versions. It is difficult to develop these functions using editorjs. I gave up

How to recover from one version to another is very difficult

**Comment by marcoancona** [comment]
My idea was to take advantage of the block nature of the underlying data and to prevent editing a block that is being edited by somebody else, by setting it read-only. Conflict resolution should be easier under the constraint that within the same block there should be no conflicts. Undo only for the changes within the block being edited. This should be possible now that blocks have a unique id. On the other hand, for this to work it should be possible for all other blocks to be updated without disrupting the editing experience of the block the user is working on.

**Reply by KartikYZ** [reply]
Hi, would you able to provide more detail into how you went about this? I am currently trying to implement the exact same functionality but can't figure out how to set certain blocks to readOnly.

**Comment by milu2003** [comment]
As for what you said about locking a block for editing, I think the user experience is very bad

At the beginning of my project, I used editor to develop it. Later, I didn't support collaboration well enough. Later, I used prosemirror to reconstruct the project. I found that prosemirror did a good job in this aspect. Editor has been abandoned by me.

You can also try tiptap, based on prosemirror：
https://www.tiptap.dev/examples/collaborative-editing#example

I think if real-time collaboration is an important feature of your product, then editorjs is not a better choice, because  the design didn't take this into account at the beginning.

**Comment by hata6502** [comment]
This issue may be related to... https://github.com/codex-team/editor.js/pull/1606

**Comment by marcoancona** [comment]
Tiptap seems interesting, thanks for the hint @milu2003. Although, I have to say, that I would prefer a block-based editor to build something closer to Notion/Confluence. And editor.js looks good out of the box.

@hata6502 indeed that would help

**Comment by milu2003** [comment]
Block building is not difficult. My current product is also in the form of block, similar to notion, but it's very easy for me to build with prosemirror, just a little more work. But the high degree of customization of prosemirror made it easier for me to build programs. What you said out of the box is both good and bad. Although it's easy to use, once you want more complex functions, I think your program will be greatly limited.

@marcoancona

**Comment by wmonecke** [comment]
Have you guys looked at this issue? #1155 Looks promising!

**Comment by SavoInfostream** [comment]
Any progress update for this feature?

**Comment by neSpecc** [comment]
> Any progress update for this feature?

Yes, we've designed and started implementing it. Here is [the first part](https://github.com/codex-team/editor.js/pull/1801). 

You can support us and boost development speed via [donation](http://opencollective.com/editorjs) on OpenCollective.

**Comment by SavoInfostream** [comment]
Donated.
Keep it going :)

**Comment by ChasLui** [comment]
Hi, Friends, we have implemented collaborative documentation (including collaborative cursors) using editor.js + yjs and we are now live. Thanks again to editor.js.

https://lanhuapp.com/dashboard/#/item

<img width="1682" alt="image" src="https://user-images.githubusercontent.com/10083758/212805242-d014aa36-eac0-46cb-8c02-4bbd73b44169.png">

<img width="529" alt="image" src="https://user-images.githubusercontent.com/10083758/212805618-0197effe-eb21-4da5-a6ad-ecde0e750db1.png">

**Reply by ahetawal-p** [reply]
Hey @ChasLui do you have some guidelines, on how you went about implementing this ?

**Reply by ChasLui** [reply]
@ahetawal-p Hi friend. You can refer to this project: [hughfenghen/y-editorjs](https://github.com/hughfenghen/y-editorjs)

**Reply by ahetawal-p** [reply]
Thank you so much @ChasLui

**Reply by ahetawal-p** [reply]
@ChasLui can you please confirm if which version of editorjs you are using in the above repo. Is this your own fork with some custom code in it to make it for collaboration ?

**Reply by ChasLui** [reply]
@ahetawal-p y-editorjs is not my code. But my implementation idea is similar to y-editorjs project.

**Reply by ZhouYC404** [reply]
Hello, I am looking for an editor with collaborative editing capabilities. Do you think editor.js+yjs is still the best solution so far (2025)?

**Reply by ChasLui** [reply]
> Hello, I am looking for an editor with collaborative editing capabilities. Do you think editor.js+yjs is still the best solution so far (2025)?

Hello, this plan has been running online for 2 years and has not encountered any major problems.

**Comment by lijie1129** [comment]
Hi,everyone.

About collaboration scenarios.

Recently, I tried to make a binding between the [YJS](https://github.com/yjs/yjs) and EditorJS.

It is to realize an idea. Please refer to it [here](https://discuss.yjs.dev/t/editor-js-bindings/1349/5?u=jarone).

I think the EditorJS(v2.x) has some issues with compatibility design for the collaboration scenarios.

## 1. Absent the reactive & incremental state of EditorJS.

The remote to local: we can use the [data](https://editorjs.io/configuration/) of EditorJS’ configuration to initialize existing content.

But the incremental updates from other remote clients to local will only use the update method of the block’ API. (EditorJS v2.x doesn’t have the update method of the block’ API, and I created it in this [PR](https://github.com/codex-team/editor.js/pull/2234))

The local updates to remote: we can listen to the EditorJS via [onChange](https://editorjs.io/configuration/) and use the update method previously of the block’s API.

## 2. Too many unnecessary DOM mutations in the EditorJS’s internal.

Update the block: please see [here](https://github.com/codex-team/editor.js/issues/2221).

Switch the read-only mode: please see [here](https://github.com/codex-team/editor.js/issues/2265).

**Reply by gohabereg** [reply]
Hi @lijie1129 

Thank you for your research! We were discussing exactly the same problems internally. Therefore we are working on the architecture redesign to enable reactive data-first approach.

That is a challenging problem as existing solutions usually work with known in advance set of content blocks. Whereas Editor.js core doesn't know much about connected Tools. We are on the research stage now, hopefully we will come up to the public with the new design soon

**Reply by SalesforceBobLightning** [reply]
This will be an amazing enhancement

**Comment by ghost** [comment]
Does anyone have a recommendation for editor that very well supports showing the difference between two versions of a file. The editor should have its own format and should be able to receive data from two sources and the create a diff. I've been searching for this for a long time now

**Comment by w99910** [comment]
https://github.com/codex-team/editor.js/assets/65523475/1eb4b789-7923-4953-8036-17ea843a2c1a

Hey guys, I just implemented responsive real-time caret and text selection which would be a feature of real-time collaboration mechanism. 

It is easy to calculate the position in fixed window size but different browsers could have different window size and font size. 

So I took current font size, line height and word spacing of the selected element into account, and pre-render lines and calculate the positions. 

I haven't refactored my code yet. If you guys are interested, I will try to release it. 😃

**Reply by w99910** [reply]
I just released a package called "editorjs-caret-selection" for caret selection feature and you can find it [here](https://github.com/w99910/editorjs-caret-selection) if you are interested. It is an initial version and there would be probably bugs. Feel free to fork it, submit pull request or modify and use as you wish. Cheers, guys! 🎉(•̀ᴗ•́ )و

**Reply by w99910** [reply]
I wrote an article about Realtime Collaborative Editor Approach using Socket.io [here](https://medium.com/@thomasbrillion/build-your-own-collaborative-realtime-notion-11c361fb2cbe). I hope I can help you all a bit.  :)

**Comment by onexdata** [comment]
Here is a dead-simple implementation of a real-time collaborative editor.js interface using OT (Operational Transforms) for conflict resolution, which is faster than CRDT-based solutions like Ys (Both Google and Microsoft opted for OT over CRDT).

I tested it up to 10 collaborators with no issues in sight. Uses ShareDB to manage the document on the server and implement OT, with no other requirements other than websocket (so you don't have to use Socket.io or other protocols).

Feel free to use it for whatever you want.

https://github.com/onexdata/realtime-editor.js

![image](https://github.com/user-attachments/assets/d0b92699-5716-4010-a688-76cea5348d57)

**Reply by DeadWisdom** [reply]
It seems like you opted to move to Yjs? Did you end up finding it simpler for this sort of application?

**Comment by bettysteger** [comment]
Has someone tried https://github.com/sebmeister2077/editorjs-realtime-collab ? 

Features:

- Realtime block add / update / move / delete
- Inline cursor & text selection visualization
- Block-level selection + pending deletion state
- **Block locking** prevents concurrent edits to the same block
- Works with any socket implementation
- Throttled updates for performance



I am just checking a few options out... Are there any other ideas how to achieve real-time collaboration with Editor.js in 2026 ?

**Comment by neSpecc** [comment]
Hello, everyone.

This feature will be available in Editor.js 3.0 which is under active development. You can watch the progress here: 
https://github.com/editor-js/document-model

How you can help:
1. Subscribe to donation https://opencollective.com/editorjs
2. Join Pull Request reviews and Discussions 
3. Become a maintainer. Contact me if you're interested, and we'll discuss your onboarding.

**Reply by bettysteger** [reply]
is there a timeline, when will this be available? how can i contact you?

**Reply by neSpecc** [reply]
Write me in Telegram: https://t.me/specc
